### DISCUSSION TITLE
Is it necessary to utilize GPU to accelerate prover ?

### DISCUSSION THREAD

**Discussion by qct** [Consensys/gnark#496] [General]
Since there are many FFT & MSM operations in Groth16, can we expect a faster prover when using GPU?

If the answer is yes, what should we do to integrate GPU prover?

**Comment by gbotrel** [comment]
hi -- I think the answer will depends on the context. If the circuit is small, CPU will do just fine. If the circuit is large, memory available on common GPUs will be an issue. 
An hybrid CPU/GPU prover would certainly be interesting (delegate some of the MSM or FFT to a GPU); having Go bindings to a library like  https://github.com/supranational/sppark would help experiment.

That being said, `gnark` performs quite well on CPU, and this has couple of advantages; easier to iterate on, easier to maintain the code (less code) and one code base runs on many targets (android/iOS, x86, arm64, WASM).

**Reply by qct** [reply]
Thanks for your answer.
I agree with you, an hybrid solution would be feasible, especially for a zkEVM project on production and want to scale TPS, circuits might be very large, speed up prover time could be an important point.

Do you have any dev plan on GPU topic? 
or just like you said, `gnark` performs well enough on CPU, it's not necessary to integrate with GPU?

**Reply by gbotrel** [reply]
We don't have "immediate/urgent" dev plan on the topic; for the current state of things, indeed the CPU perf is sufficient.
